core/node/node_framework/src/implementations/layers/state_keeper/output_handler.rs

```diff
@@ -84,6 +89,38 @@ impl OutputHandlerLayer {
         self.protective_reads_persistence_enabled = protective_reads_persistence_enabled;
         self
     }
+
+    async fn validate_l2_legacy_shared_bridge_addr(
+        &self,
+        pool: &ConnectionPool<Core>,
+    ) -> Result<(), WiringError> {
+        let mut connection = pool.connection().await.context("Get DB connection")?;
+
+        if let Some(l2_block) = connection
+            .blocks_dal()
+            .get_earliest_l2_block_number()
+            .await
+            .context("failed to load earliest l2 block number")?
+        {
+            let header = connection
+                .blocks_dal()
+                .get_l2_block_header(l2_block)
+                .await
+                .context("failed to load L2 block header")?
+                .context("missing L2 block header")?;
+            let protocol_version = header
+                .protocol_version
+                .unwrap_or_else(ProtocolVersionId::last_potentially_undefined);
+
+            if protocol_version.is_pre_gateway() && self.l2_legacy_shared_bridge_addr.is_none() {
```

Comment:
I don't think this check will work with pruning enabled (i.e., it can eventually lead to a node panicking). Maybe, it makes sense to query the protocol version of the genesis block instead, and skip the check if there's no genesis (i.e., a node was recovered from a snapshot and/or has pruning enabled)? Or will such a check be not strict enough?

Reply:
if info about earliest batches is missing then check will still work correctly (but won't be that strict) because `earliest_batch.protocol_version >= genesis_batch.protocol_version` and `is_pre_gateway` is monotonic. Also, check will be performed for main nodes that always have full history and ENs fetch data from main node so transitively it means data they receive is already strict-checked